Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59380 Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose**

Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

fixes https://github.com/pytorch/pytorch/issues/64176

anjali411 committed:
Add fast path for addmm when the inputs are conjugate

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"




[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"




[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"




[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28898374).

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1031,13 +1031,22 @@ static void addmm_impl_cpu_(
   const int64_t ldc = c.strides()[transpose_c ? 0 : 1];
 
+  if (a.is_conj() && !transpose_a) {
+    a.conj_physical_();
```
This potentially modifies an input tensor's values.

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


TODO: test performance and remove some boilerplate fallback code

Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 commented:
<details open>
<summary>Benchmarks (about 10-20% (and ~50% in many cases) improvement on CPU and (about 1-2% (and 20-50% for some sizes) on CUDA:</summary>

[---------- addmm input dtype:torch.complex128 device:cpu -----------]
                       |  Using conj_physical  |  Using conjugate view
1 threads: -----------------------------------------------------------
      (2, 2, 2)        |           12.4        |           12.7       
      (2, 2, 8)        |           13.6        |           13.9       
      (2, 2, 64)       |           23.2        |           21.1       
      (2, 2, 128)      |           24.6        |           23.1       
      (2, 2, 256)      |           28.4        |           25.1       
      (2, 2, 512)      |           23.9        |           22.1       
      (2, 8, 2)        |           14.3        |           12.7       
      (2, 8, 8)        |           14.9        |           14.5       
      (2, 8, 64)       |           16.8        |           15.3       
      (2, 8, 128)      |           18.2        |           17.4       
      (2, 8, 256)      |           25.1        |           22.3       
      (2, 8, 512)      |           32.0        |           28.3       
      (2, 64, 2)       |           14.2        |           12.9       
      (2, 64, 8)       |           15.7        |           15.7       
      (2, 64, 64)      |           25.9        |           21.4       
      (2, 64, 128)     |           38.4        |           26.6       
      (2, 64, 256)     |           71.4        |           50.5       
      (2, 64, 512)     |          118.7        |           70.0       
      (2, 128, 2)      |           21.7        |           13.5       
      (2, 128, 8)      |           17.5        |           16.7       
      (2, 128, 64)     |           38.0        |           26.4       
      (2, 128, 128)    |           59.1        |           36.3       
      (2, 128, 256)    |           97.3        |           53.5       
      (2, 128, 512)    |          179.5        |           88.6       
      (2, 256, 2)      |           14.6        |           14.4       
      (2, 256, 8)      |           22.0        |           19.2       
      (2, 256, 64)     |           65.0        |           37.9       
      (2, 256, 128)    |          103.1        |           58.0       
      (2, 256, 256)    |          194.1        |          102.2       
      (2, 256, 512)    |          362.9        |          185.4       
      (2, 512, 2)      |           16.6        |           16.0       
      (2, 512, 8)      |           28.4        |           23.8       
      (2, 512, 64)     |          104.5        |           59.6       
      (2, 512, 128)    |          189.1        |           96.5       
      (2, 512, 256)    |          346.6        |          173.4       
      (2, 512, 512)    |          690.2        |          340.0       
      (8, 2, 2)        |           13.0        |           13.4       
      (8, 2, 8)        |           14.7        |           14.8       
      (8, 2, 64)       |           16.5        |           16.0       
      (8, 2, 128)      |           18.9        |           18.1       
      (8, 2, 256)      |           26.9        |           24.2       
      (8, 2, 512)      |           49.5        |           42.0       
      (8, 8, 2)        |           12.9        |           13.0       
      (8, 8, 8)        |           13.8        |           13.6       
      (8, 8, 64)       |           18.3        |           17.8       
      (8, 8, 128)      |           22.6        |           20.4       
      (8, 8, 256)      |           32.9        |           28.7       
      (8, 8, 512)      |           64.6        |           52.2       
      (8, 64, 2)       |           14.3        |           14.4       
      (8, 64, 8)       |           18.0        |           17.0       
      (8, 64, 64)      |           33.2        |           27.4       
      (8, 64, 128)     |           50.6        |           38.2       
      (8, 64, 256)     |           85.3        |           62.5       
      (8, 64, 512)     |          164.5        |          115.3       
      (8, 128, 2)      |           21.3        |           15.7       
      (8, 128, 8)      |           21.3        |           19.6       
      (8, 128, 64)     |           49.7        |           37.3       
      (8, 128, 128)    |           79.7        |           56.0       
      (8, 128, 256)    |          139.7        |           96.6       
      (8, 128, 512)    |          275.6        |          180.7       
      (8, 256, 2)      |           18.0        |           18.1       
      (8, 256, 8)      |           28.3        |           25.9       
      (8, 256, 64)     |           84.3        |           62.2       
      (8, 256, 128)    |          147.0        |          100.2       
      (8, 256, 256)    |          278.9        |          210.3       
      (8, 256, 512)    |          548.7        |          363.5       
      (8, 512, 2)      |           25.1        |           29.1       
      (8, 512, 8)      |           43.3        |           37.9       
      (8, 512, 64)     |          255.8        |          164.9       
      (8, 512, 128)    |          276.6        |          182.9       
      (8, 512, 256)    |          526.1        |          341.8       
      (8, 512, 512)    |         1018.9        |          659.1       
      (64, 2, 2)       |           14.6        |           14.0       
      (64, 2, 8)       |           25.9        |           16.4       
      (64, 2, 64)      |           45.3        |           38.9       
      (64, 2, 128)     |           73.7        |           61.5       
      (64, 2, 256)     |          132.8        |          104.4       
      (64, 2, 512)     |          233.3        |          188.4       
      (64, 8, 2)       |           15.3        |           15.1       
      (64, 8, 8)       |           18.4        |           21.2       
      (64, 8, 64)      |           51.4        |           48.2       
      (64, 8, 128)     |           84.4        |           73.8       
      (64, 8, 256)     |          154.6        |          128.7       
      (64, 8, 512)     |          284.8        |          242.1       
      (64, 64, 2)      |           24.3        |           24.6       
      (64, 64, 8)      |           37.2        |           35.2       
      (64, 64, 64)     |          115.3        |           99.2       
      (64, 64, 128)    |          196.7        |          169.7       
      (64, 64, 256)    |          368.1        |          319.6       
      (64, 64, 512)    |          687.3        |          597.0       
      (64, 128, 2)     |           36.4        |           36.3       
      (64, 128, 8)     |           58.8        |           53.8       
      (64, 128, 64)    |          183.1        |          169.1       
      (64, 128, 128)   |          536.3        |          295.2       
      (64, 128, 256)   |          611.8        |          545.7       
      (64, 128, 512)   |         1147.9        |         1015.8       
      (64, 256, 2)     |           56.6        |           58.4       
      (64, 256, 8)     |           98.4        |           94.5       
      (64, 256, 64)    |          333.8        |          306.4       
      (64, 256, 128)   |          585.3        |          531.2       
      (64, 256, 256)   |         1077.6        |          987.4       
      (64, 256, 512)   |         2112.3        |         1902.8       
      (64, 512, 2)     |           94.7        |           96.2       
      (64, 512, 8)     |          172.5        |          177.8       
      (64, 512, 64)    |          616.0        |          563.1       
      (64, 512, 128)   |         1088.2        |         1044.5       
      (64, 512, 256)   |         2103.6        |         1912.0       
      (64, 512, 512)   |         4105.4        |         3719.3       
      (128, 2, 2)      |           16.3        |           15.6       
      (128, 2, 8)      |           18.6        |           18.3       
      (128, 2, 64)     |           69.8        |           61.2       
      (128, 2, 128)    |          120.4        |          100.3       
      (128, 2, 256)    |          235.8        |          191.6       
      (128, 2, 512)    |          463.6        |          363.6       
      (128, 8, 2)      |           20.6        |           17.5       
      (128, 8, 8)      |           23.4        |           22.2       
      (128, 8, 64)     |           85.4        |           70.5       
      (128, 8, 128)    |          161.5        |          132.6       
      (128, 8, 256)    |          282.4        |          234.2       
      (128, 8, 512)    |          579.9        |          440.3       
      (128, 64, 2)     |           42.4        |           36.7       
      (128, 64, 8)     |           61.4        |           55.4       
      (128, 64, 64)    |          197.3        |          180.1       
      (128, 64, 128)   |          349.7        |          315.8       
      (128, 64, 256)   |          670.7        |          599.2       
      (128, 64, 512)   |         1313.5        |         1183.8       
      (128, 128, 2)    |           72.5        |           71.5       
      (128, 128, 8)    |           98.3        |           95.1       
      (128, 128, 64)   |          335.9        |          310.5       
      (128, 128, 128)  |          591.3        |          549.6       
      (128, 128, 256)  |         1121.5        |         1039.5       
      (128, 128, 512)  |         2198.2        |         2577.7       
      (128, 256, 2)    |           97.7        |           99.8       
      (128, 256, 8)    |          191.0        |          171.8       
      (128, 256, 64)   |          596.3        |          568.6       
      (128, 256, 128)  |         1154.4        |         1016.3       
      (128, 256, 256)  |         2046.4        |         1946.0       
      (128, 256, 512)  |         4179.5        |         3751.3       
      (128, 512, 2)    |          189.7        |          175.0       
      (128, 512, 8)    |          324.4        |          314.0       
      (128, 512, 64)   |         1209.7        |         1070.8       
      (128, 512, 128)  |         2048.8        |         1949.7       
      (128, 512, 256)  |         3938.4        |         3703.0       
      (128, 512, 512)  |         8362.2        |         7391.9       
      (256, 2, 2)      |           19.8        |           20.7       
      (256, 2, 8)      |           25.3        |           23.1       
      (256, 2, 64)     |          125.2        |          100.9       
      (256, 2, 128)    |          229.9        |          193.8       
      (256, 2, 256)    |          448.9        |          361.7       
      (256, 2, 512)    |          864.8        |          692.8       
      (256, 8, 2)      |           22.9        |           22.2       
      (256, 8, 8)      |           32.5        |           31.3       
      (256, 8, 64)     |          143.8        |          120.5       
      (256, 8, 128)    |          264.4        |          218.0       
      (256, 8, 256)    |          525.7        |          440.6       
      (256, 8, 512)    |         1021.5        |          844.2       
      (256, 64, 2)     |           61.4        |           63.9       
      (256, 64, 8)     |           97.1        |           92.2       
      (256, 64, 64)    |          364.0        |          334.7       
      (256, 64, 128)   |          663.3        |          607.0       
      (256, 64, 256)   |         1289.1        |         1175.1       
      (256, 64, 512)   |         2501.1        |         2275.9       
      (256, 128, 2)    |          100.7        |          100.2       
      (256, 128, 8)    |          171.3        |          171.1       
      (256, 128, 64)   |          613.9        |          579.3       
      (256, 128, 128)  |         1142.2        |         1067.4       
      (256, 128, 256)  |         2170.2        |         2026.8       
      (256, 128, 512)  |         4206.8        |         3933.0       
      (256, 256, 2)    |          193.1        |          193.0       
      (256, 256, 8)    |          320.5        |          316.9       
      (256, 256, 64)   |         1142.0        |         1093.1       
      (256, 256, 128)  |         2045.6        |         1970.5       
      (256, 256, 256)  |         3930.4        |         3746.0       
      (256, 256, 512)  |         7601.9        |         7273.3       
      (256, 512, 2)    |          320.6        |          337.4       
      (256, 512, 8)    |          614.1        |          600.7       
      (256, 512, 64)   |         2124.7        |         2161.4       
      (256, 512, 128)  |         3927.8        |         3752.1       
      (256, 512, 256)  |         7486.5        |         7667.3       
      (256, 512, 512)  |        15605.3        |        14979.7       
      (512, 2, 2)      |           26.9        |           24.5       
      (512, 2, 8)      |           48.7        |           43.1       
      (512, 2, 64)     |          230.9        |          185.4       
      (512, 2, 128)    |          441.3        |          349.0       
      (512, 2, 256)    |          873.3        |          703.1       
      (512, 2, 512)    |         1858.5        |         1411.1       
      (512, 8, 2)      |           35.8        |           32.3       
      (512, 8, 8)      |           62.1        |           53.8       
      (512, 8, 64)     |          267.2        |          216.4       
      (512, 8, 128)    |          501.3        |          411.0       
      (512, 8, 256)    |         1065.6        |          877.2       
      (512, 8, 512)    |         2104.9        |         1717.0       
      (512, 64, 2)     |          119.4        |          114.4       
      (512, 64, 8)     |          198.7        |          181.8       
      (512, 64, 64)    |          692.9        |          640.6       
      (512, 64, 128)   |         1283.5        |         1177.2       
      (512, 64, 256)   |         2499.5        |         2322.6       
      (512, 64, 512)   |         5681.5        |         4513.2       
      (512, 128, 2)    |          181.8        |          242.2       
      (512, 128, 8)    |          329.9        |          324.0       
      (512, 128, 64)   |         1175.4        |         1124.2       
      (512, 128, 128)  |         2166.1        |         2050.6       
      (512, 128, 256)  |         4172.3        |         3967.5       
      (512, 128, 512)  |         8302.4        |         7761.5       
      (512, 256, 2)    |          338.6        |          344.7       
      (512, 256, 8)    |          635.8        |          625.2       
      (512, 256, 64)   |         2351.6        |         2139.3       
      (512, 256, 128)  |         4012.5        |         3853.1       
      (512, 256, 256)  |         7624.4        |         7325.6       
      (512, 256, 512)  |        15684.2        |        14566.9       
      (512, 512, 2)    |          627.0        |          629.2       
      (512, 512, 8)    |         1180.2        |         1150.9       
      (512, 512, 64)   |         4112.5        |         4037.9       
      (512, 512, 128)  |         7583.2        |         7366.5       
      (512, 512, 256)  |        14937.5        |        23224.9       
      (512, 512, 512)  |        30577.4        |        28196.8       

Times are in microseconds (us).

[---------- addmm input dtype:torch.complex128 device:cuda ----------]
                       |  Using conj_physical  |  Using conjugate view
1 threads: -----------------------------------------------------------
      (2, 2, 2)        |           93.1        |           44.8       
      (2, 2, 8)        |           61.2        |           44.8       
      (2, 2, 64)       |           60.7        |           43.0       
      (2, 2, 128)      |           60.2        |           43.6       
      (2, 2, 256)      |           59.8        |           43.8       
      (2, 2, 512)      |           60.7        |           43.9       
      (2, 8, 2)        |           60.7        |           44.2       
      (2, 8, 8)        |           61.6        |           44.4       
      (2, 8, 64)       |           60.7        |           43.3       
      (2, 8, 128)      |           59.0        |           40.8       
      (2, 8, 256)      |           57.4        |           41.9       
      (2, 8, 512)      |           60.5        |           44.2       
      (2, 64, 2)       |          148.2        |          143.1       
      (2, 64, 8)       |          101.6        |           96.4       
      (2, 64, 64)      |          102.6        |           97.2       
      (2, 64, 128)     |          152.8        |          146.4       
      (2, 64, 256)     |          153.3        |          146.1       
      (2, 64, 512)     |          157.3        |          147.9       
      (2, 128, 2)      |          276.5        |          271.1       
      (2, 128, 8)      |          184.3        |          178.9       
      (2, 128, 64)     |          277.7        |          272.0       
      (2, 128, 128)    |          280.7        |          273.6       
      (2, 128, 256)    |          286.0        |          276.8       
      (2, 128, 512)    |          290.1        |          277.1       
      (2, 256, 2)      |          531.7        |          526.4       
      (2, 256, 8)      |          349.6        |          344.1       
      (2, 256, 64)     |          533.4        |          526.9       
      (2, 256, 128)    |          543.8        |          534.9       
      (2, 256, 256)    |          548.6        |          535.7       
      (2, 256, 512)    |          562.5        |          535.8       
      (2, 512, 2)      |         1043.3        |         1037.6       
      (2, 512, 8)      |         1043.3        |         1038.0       
      (2, 512, 64)     |         1046.9        |         1039.0       
      (2, 512, 128)    |         1053.7        |         1040.7       
      (2, 512, 256)    |         1072.4        |         1045.7       
      (2, 512, 512)    |         1111.6        |         1057.2       
      (8, 2, 2)        |           60.3        |           42.0       
      (8, 2, 8)        |           65.4        |           43.3       
      (8, 2, 64)       |           59.1        |           58.6       
      (8, 2, 128)      |           86.4        |           43.2       
      (8, 2, 256)      |           59.8        |           43.4       
      (8, 2, 512)      |           61.0        |           44.1       
      (8, 8, 2)        |           60.1        |           43.7       
      (8, 8, 8)        |           60.7        |           43.2       
      (8, 8, 64)       |           64.4        |           44.2       
      (8, 8, 128)      |           59.9        |           42.6       
      (8, 8, 256)      |           60.9        |           43.5       
      (8, 8, 512)      |           60.7        |           44.0       
      (8, 64, 2)       |          148.4        |          143.3       
      (8, 64, 8)       |          101.6        |           96.4       
      (8, 64, 64)      |          105.8        |           99.6       
      (8, 64, 128)     |          154.7        |          148.2       
      (8, 64, 256)     |          155.8        |          148.4       
      (8, 64, 512)     |          157.6        |          148.0       
      (8, 128, 2)      |          276.4        |          271.0       
      (8, 128, 8)      |          184.6        |          179.2       
      (8, 128, 64)     |          280.9        |          274.3       
      (8, 128, 128)    |          284.3        |          277.2       
      (8, 128, 256)    |          287.1        |          277.6       
      (8, 128, 512)    |          290.5        |          277.3       
      (8, 256, 2)      |          349.6        |          344.3       
      (8, 256, 8)      |          532.1        |          526.5       
      (8, 256, 64)     |          536.8        |          529.6       
      (8, 256, 128)    |          539.0        |          530.0       
      (8, 256, 256)    |          543.5        |          530.8       
      (8, 256, 512)    |          563.6        |          536.2       
      (8, 512, 2)      |          680.0        |          674.5       
      (8, 512, 8)      |         1043.9        |         1037.9       
      (8, 512, 64)     |         1049.9        |         1040.4       
      (8, 512, 128)    |         1054.4        |         1041.6       
      (8, 512, 256)    |         1080.4        |         1052.9       
      (8, 512, 512)    |         1112.3        |         1056.4       
      (64, 2, 2)       |           56.7        |           41.0       
      (64, 2, 8)       |           85.6        |           59.7       
      (64, 2, 64)      |          106.5        |           42.9       
      (64, 2, 128)     |           62.1        |           43.8       
      (64, 2, 256)     |           59.6        |           42.7       
      (64, 2, 512)     |           61.0        |           42.9       
      (64, 8, 2)       |           60.4        |           43.9       
      (64, 8, 8)       |           60.5        |           41.6       
      (64, 8, 64)      |           58.4        |           43.0       
      (64, 8, 128)     |           59.3        |           43.1       
      (64, 8, 256)     |           58.1        |           41.0       
      (64, 8, 512)     |           57.4        |           41.7       
      (64, 64, 2)      |          149.4        |          144.5       
      (64, 64, 8)      |          152.7        |          146.7       
      (64, 64, 64)     |          154.5        |          148.4       
      (64, 64, 128)    |          155.7        |          148.7       
      (64, 64, 256)    |          157.6        |          149.3       
      (64, 64, 512)    |          166.8        |          151.2       
      (64, 128, 2)     |          277.9        |          272.6       
      (64, 128, 8)     |          280.8        |          274.9       
      (64, 128, 64)    |          281.8        |          275.0       
      (64, 128, 128)   |          286.0        |          278.3       
      (64, 128, 256)   |          289.1        |          278.8       
      (64, 128, 512)   |          305.2        |          283.0       
      (64, 256, 2)     |          351.6        |          346.4       
      (64, 256, 8)     |          536.9        |          530.9       
      (64, 256, 64)    |          538.5        |          531.3       
      (64, 256, 128)   |          544.4        |          534.6       
      (64, 256, 256)   |          557.6        |          538.3       
      (64, 256, 512)   |          580.5        |          547.2       
      (64, 512, 2)     |         1047.1        |         1041.5       
      (64, 512, 8)     |         1050.1        |         1043.8       
      (64, 512, 64)    |         1066.0        |         1056.3       
      (64, 512, 128)   |         1075.1        |         1056.5       
      (64, 512, 256)   |         1090.6        |         1060.9       
      (64, 512, 512)   |         1138.1        |         1071.7       
      (128, 2, 2)      |           57.5        |           42.1       
      (128, 2, 8)      |           59.0        |           43.7       
      (128, 2, 64)     |           61.5        |           42.8       
      (128, 2, 128)    |           59.7        |           44.4       
      (128, 2, 256)    |           59.9        |           43.1       
      (128, 2, 512)    |           79.4        |           66.2       
      (128, 8, 2)      |           59.0        |           46.0       
      (128, 8, 8)      |           61.4        |           43.8       
      (128, 8, 64)     |           60.1        |           43.1       
      (128, 8, 128)    |           61.1        |           44.7       
      (128, 8, 256)    |           62.1        |           43.2       
      (128, 8, 512)    |           83.1        |           66.3       
      (128, 64, 2)     |          152.5        |          146.8       
      (128, 64, 8)     |          154.9        |          148.7       
      (128, 64, 64)    |          155.6        |          149.0       
      (128, 64, 128)   |          157.4        |          149.5       
      (128, 64, 256)   |          163.7        |          151.4       
      (128, 64, 512)   |          313.7        |          294.2       
      (128, 128, 2)    |          283.8        |          278.1       
      (128, 128, 8)    |          284.8        |          278.6       
      (128, 128, 64)   |          284.4        |          277.7       
      (128, 128, 128)  |          288.4        |          279.8       
      (128, 128, 256)  |          300.3        |          283.3       
      (128, 128, 512)  |          579.1        |          554.6       
      (128, 256, 2)    |          544.5        |          538.4       
      (128, 256, 8)    |          544.8        |          538.9       
      (128, 256, 64)   |          546.7        |          539.1       
      (128, 256, 128)  |          553.0        |          540.9       
      (128, 256, 256)  |          570.7        |          549.5       
      (128, 256, 512)  |         1114.0        |         1077.1       
      (128, 512, 2)    |         1054.3        |         1047.6       
      (128, 512, 8)    |         1065.7        |         1060.1       
      (128, 512, 64)   |         1075.7        |         1062.0       
      (128, 512, 128)  |         1082.3        |         1066.0       
      (128, 512, 256)  |         1107.3        |         1073.2       
      (128, 512, 512)  |         2186.8        |         2114.5       
      (256, 2, 2)      |           57.4        |           44.0       
      (256, 2, 8)      |           59.9        |           43.4       
      (256, 2, 64)     |           60.6        |           42.5       
      (256, 2, 128)    |           61.5        |           44.6       
      (256, 2, 256)    |           79.3        |           66.1       
      (256, 2, 512)    |          175.7        |          149.6       
      (256, 8, 2)      |           60.7        |           43.6       
      (256, 8, 8)      |           60.9        |           44.1       
      (256, 8, 64)     |           61.3        |           43.7       
      (256, 8, 128)    |           61.3        |           44.0       
      (256, 8, 256)    |           80.0        |           66.2       
      (256, 8, 512)    |          178.0        |          151.2       
      (256, 64, 2)     |          154.3        |          148.5       
      (256, 64, 8)     |          155.3        |          149.1       
      (256, 64, 64)    |          157.5        |          150.3       
      (256, 64, 128)   |          164.2        |          152.7       
      (256, 64, 256)   |          311.9        |          295.9       
      (256, 64, 512)   |          637.7        |          605.2       
      (256, 128, 2)    |          286.3        |          280.3       
      (256, 128, 8)    |          287.0        |          280.9       
      (256, 128, 64)   |          289.1        |          281.4       
      (256, 128, 128)  |          300.0        |          286.0       
      (256, 128, 256)  |          576.8        |          558.1       
      (256, 128, 512)  |         1164.1        |         1126.7       
      (256, 256, 2)    |          548.5        |          542.4       
      (256, 256, 8)    |          549.2        |          542.9       
      (256, 256, 64)   |          556.5        |          544.9       
      (256, 256, 128)  |          570.7        |          558.0       
      (256, 256, 256)  |         1105.2        |         1082.3       
      (256, 256, 512)  |         2210.6        |         2159.6       
      (256, 512, 2)    |         1068.6        |         1061.2       
      (256, 512, 8)    |         1081.1        |         1073.0       
      (256, 512, 64)   |         1091.2        |         1079.5       
      (256, 512, 128)  |         1110.4        |         1091.6       
      (256, 512, 256)  |         2170.5        |         2133.1       
      (256, 512, 512)  |         4322.5        |         4242.7       
      (512, 2, 2)      |           61.4        |           44.6       
      (512, 2, 8)      |           61.6        |           44.4       
      (512, 2, 64)     |           60.4        |           44.1       
      (512, 2, 128)    |           79.8        |           68.1       
      (512, 2, 256)    |          176.7        |          149.7       
      (512, 2, 512)    |          342.5        |          283.8       
      (512, 8, 2)      |           59.3        |           47.4       
      (512, 8, 8)      |           60.6        |           44.1       
      (512, 8, 64)     |           60.7        |           43.5       
      (512, 8, 128)    |           80.5        |           68.9       
      (512, 8, 256)    |          178.5        |          151.3       
      (512, 8, 512)    |          344.3        |          285.2       
      (512, 64, 2)     |          157.0        |          151.0       
      (512, 64, 8)     |          158.0        |          151.6       
      (512, 64, 64)    |          167.1        |          155.4       
      (512, 64, 128)   |          315.3        |          302.3       
      (512, 64, 256)   |          637.4        |          607.1       
      (512, 64, 512)   |         1028.6        |          966.1       
      (512, 128, 2)    |          290.4        |          284.2       
      (512, 128, 8)    |          290.7        |          284.4       
      (512, 128, 64)   |          305.6        |          293.4       
      (512, 128, 128)  |          582.3        |          565.5       
      (512, 128, 256)  |         1164.1        |         1131.0       
      (512, 128, 512)  |         1808.8        |         1743.5       
      (512, 256, 2)    |          562.8        |          555.6       
      (512, 256, 8)    |          564.2        |          556.2       
      (512, 256, 64)   |          580.3        |          568.4       
      (512, 256, 128)  |         1115.4        |         1098.1       
      (512, 256, 256)  |         2214.3        |         2179.7       
      (512, 256, 512)  |         3384.2        |         3304.9       
      (512, 512, 2)    |         1110.7        |         1104.3       
      (512, 512, 8)    |         1112.2        |         1104.8       
      (512, 512, 64)   |         1137.3        |         1121.7       
      (512, 512, 128)  |         2191.5        |         2167.3       
      (512, 512, 256)  |         4322.8        |         4275.0       
      (512, 512, 512)  |         6546.1        |         6437.2       

Times are in microseconds (us).

</details>

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


TODO: test performance and remove some boilerplate fallback code

Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -9,6 +9,39 @@ namespace at { namespace native {
+    return resolve_conj_if_indicated(tensor, (transpose_tensor == transpose_result));
+  } else {
+    transpose_tensor = true;
+    if (tensor.is_conj() && (transpose_tensor != transpose_result)) {
```
as for the cpu, this special case is probably not needed, and you can always rely on `clone` to resolve conj.

ngimel reviewed (commented):
This looks good, I have questions about logic, also can you please benchmark instruction counts (similar to #55107)

lezcano reviewed (commented):
I believe it is possible to always compute this operation without having to copy the matrices (now we copy at times via `resolve_conj` or `clone`). To do this, note that the BLAS kernels compute `op(A)op(B)` where `op \in {Norma, Transpose, Conjugate Transpose}` (`n`, ` t`, `c`). As such, the only cases that are problematic is when one or both of the `op' s are not transposed but are conjugated.

If both of `op` are conjugated, we can compute `AB` passing `n` mode if they were not transposed and `t` if they were and then return the result conjugated. This just performs the transformation `AB = C` => `conj(A)conj(B) = conj(C)`.

After performing the previous transformation if necessary, at most one of `A` or `B` is conjugated. Assume wlog that it's `A`. Assume also that it's not transposed, otherwise we could  pass it to the kernel. Assume that `B` is not transposed (the transposed case is equivalent). Then, we can compute compute `BA` with `B` in mode `t` and `A` in mode `c` and then return the result transposed.

anjali411 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1034,27 +1038,36 @@ static void addmm_impl_cpu_(
   const int64_t ldc = c.strides()[transpose_c ? 0 : 1];
 
+  // Always ensure the conjugation for c is resolved since there's no way to specify c's conjugation in the gemm call
+  TORCH_CHECK(!c.is_conj());
```
yeah makes sense. will change it to `TORCH_INTERNAL_ASSERT_DEBUG_ONLY`

anjali411 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1017,14 +1016,19 @@ static void addmm_impl_cpu_(
     transpose_a = !transpose_c;
-    a = m1.clone(at::MemoryFormat::Contiguous);
+    if (m1.is_conj() && transpose_a) {
+      a = m1.conj().clone(at::MemoryFormat::Contiguous);
```
why do you not expect any performance gain by using `ConjTranspose` here? is it because the conjugation is resolved in clone and so there's no additional kernel call?

ngimel replied:
Yes. You have to do `.clone` either way, and then, whether you send `Transpose` or `ConjTranspose` to blas, blas performance itself doesn't change. So this special casing doesn't make things worse, but it also doesn't make them better and adds code complexity.

anjali411 commented:
> After performing the previous transformation if necessary, at most one of `A` or `B` is conjugated. Assume wlog that it's `A`. Assume also that it's not transposed, otherwise we could pass it to the kernel. Assume that `B` is not transposed (the transposed case is equivalent). Then, we can compute compute `BA` with `B` in mode `t` and `A` in mode `c` and then return the result transposed.

Is it ok to return a non-contiguous tensor when the inputs are contiguous? Do we have a similar behavior for any other linear algebra operations?

ezyang commented:
@anjali411 is right, we don't really have any precedent returning transposed inputs from non-transposed inputs; our typical invariant is that you can assume the return of an operator that does actual computation is contiguous (or preserves the input layout).

That being said, it's pretty attractive idea of optimizing all of these cases, and there are not too many BC considerations because this is specifically for complex matrices.

Lezcano commented:
Most linalg operations return Fortran-contig tensors. We used to have note in the docs, but we removed it because it was too technical for the average user.

ezyang commented:
OK, doesn't sound like there's any downside.

ngimel commented:
That's a cool idea @lezcano, but matmul ops, up until now, have always returned contiguous tensors. And, given that matmuls are common in nn workloads, their returns are much more likely to mingle with contiguous tensors in subsequent ops, where having tensors with the different strides will probably lead to perf penalty. We could still say that as long as its limited to complex tensors that's ok, but I'm afraid it'll be confusing if output layout depends on whether conj bit is set on inputs or not.

Lezcano commented:
I see that the way this is handled at the moment is via a copy into a contiguous tensor:
https://github.com/pytorch/pytorch/blob/1354179e06600598e5d5374bff496009e9883e82/aten/src/ATen/native/LinearAlgebra.cpp#L1079-L1081

As such, this copy would make the "transpose" above (and I say "transpose" because it's a transpose of a FORTRAN-contig tensor) into a contiguous tensor, so the invariant of always returning a contiguous tensor would still hold.

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

TODO: Benchmark `torch.linalg.qr` and `torch.linalg.svd` backward functions

Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

ngimel commented:
Code changes look good, but you probably want to add sample inputs with conjugated second arg to the tests? Also, it's probably better to adjust tolerance rather than skip the conj view test?

anjali411 committed:
Update on "Add fast path for addmm when the inputs are conjugate"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

TODO: Benchmark `torch.linalg.qr` and `torch.linalg.svd` backward functions

Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

TODO: Benchmark `torch.linalg.qr` and `torch.linalg.svd` backward functions

Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

TODO: Benchmark `torch.linalg.qr` and `torch.linalg.svd` backward functions

Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

TODO: Benchmark `torch.linalg.qr` and `torch.linalg.svd` backward functions

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

ngimel reviewed (approved):
Hopefully tests pass

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

Lezcano commented:
Are the `resolve_conj` not going to be elided in the end? If so, perhaps opening an issue and reference it in the code to point out that this optimisation is possible?

ngimel commented:
They are going to be elided for mms, but not for bmms (looks like bmm in some cases is buggy on cuda, issue is opened #64103)

Lezcano commented:
The current algorithm does still make some copies via `resolve_conj` in some cases. In https://github.com/pytorch/pytorch/pull/59380#pullrequestreview-725310492 I outlined a way to dispatch the arguments so that no copy is needed as long as the input matrices are both Fortran-contig or Fortran-transposed.

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -1411,17 +1411,18 @@ Tensor from_file(c10::string_view filename, c10::optional<bool> shared, c10::opt
 Tensor clone(const Tensor& src, c10::optional<c10::MemoryFormat> optional_memory_format) {
   auto memory_format =
       optional_memory_format.value_or(MemoryFormat::Preserve);
+  Tensor self;
```
fixes https://github.com/pytorch/pytorch/issues/64176

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28898374).

ngimel commented on test/test_linalg.py:
```diff
@@ -6165,6 +6165,41 @@ def genf_float(x, y):
 
+    @onlyOnCPUAndCUDA
+    def test_mm_bmm_non_memory_dense(self, device):
+        def _slice(tensor, fn):
+            return fn(tensor)[:, ::2]
```
```suggestion
        def _slice(tensor, fn):
            return fn(tensor)[..., ::2]
```
this can be used for batched tensor also

ngimel commented on test/test_linalg.py:
```diff
@@ -6165,6 +6165,41 @@ def genf_float(x, y):
+        Ab = torch.randn(2, 3, 6, dtype=torch.cfloat, device=device)
+        Bb = torch.randn(2, 3, 3, dtype=torch.cfloat, device=device)
+        Bb_ = torch.randn(1, 3, 3, dtype=torch.cfloat, device=device).expand(2, 3, 3)
+        out_b = torch.empty(2, 3, 3, device=device, dtype=torch.complex64).transpose(-1, 1)
```
nit: `transpose(-1, -2)`, or `transpose(1,2)`, this mixing of negative and positive indices looks confusing (same in `t_b` function)

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

fixes https://github.com/pytorch/pytorch/issues/64176

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

fixes https://github.com/pytorch/pytorch/issues/64176

[ghstack-poisoned]

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

fixes https://github.com/pytorch/pytorch/issues/64176

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28898374).

anjali411 committed:
Update on "Add fast path for torch.{mm, addmm, bmm, baddbmm} when the inputs are conjugate transpose"


Benchmarks for addmm: https://github.com/pytorch/pytorch/pull/59380#issuecomment-893730946

Disabled fastpath for bmm and baddbmm, due to this bug: https://github.com/pytorch/pytorch/issues/64103

30% decrease in peak memory usage when `conj_physical` is used in the code sample below as well as the backward formulas:
```
import torch

x=torch.randn(2, 3, dtype=torch.cdouble, device='cuda', requires_grad=True).conj()
y=torch.randn(3, 2, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()
z=torch.randn(3, 3, dtype=torch.cdouble, device='cuda',requires_grad=True).transpose(0, 1).conj()

out=torch.addmm(x, y, z)

out.sum().backward()

print(torch.cuda.memory_summary())
```
Differential Revision: [D28898374](https://our.internmc.facebook.com/intern/diff/D28898374)

fixes https://github.com/pytorch/pytorch/issues/64176

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28898374).
